### PR TITLE
New routing

### PR DESCRIPTION
**Pull request by JonaVDM**
For now a draft to check stuff

### CODE REVIEW

File: __pycache__/config.cpython-310.pyc
Comment (JonaVDM) [inline __pycache__/config.cpython-310.pyc:1]: Why?

File: libs/__pycache__/__init__.cpython-310.pyc
Comment (JonaVDM) [inline libs/__pycache__/__init__.cpython-310.pyc:1]: Remove this

File: libs/controllers/config/__pycache__/Ledger.cpython-310.pyc
Comment (JonaVDM) [inline libs/controllers/config/__pycache__/Ledger.cpython-310.pyc:1]: Remove this

File: libs/controllers/config/__pycache__/NodeConfigData.cpython-310.pyc
Comment (JonaVDM) [inline libs/controllers/config/__pycache__/NodeConfigData.cpython-310.pyc:1]: Remove this

File: libs/controllers/config/__pycache__/__init__.cpython-310.pyc
Comment (JonaVDM) [inline libs/controllers/config/__pycache__/__init__.cpython-310.pyc:1]: Remove this

File: libs/Node.py [hunk 1]
```diff
@@ -71,13 +88,17 @@ def __init__(self,
...
+                                                  self.routing_controller.handle_route_response)
         self.network_controller.register_callback(-1,
                                                   self.neighbours_controller.handle_alive)
         
         self.network_controller.register_callback(Frame.FRAME_TYPES['sync_time'],
                                                     lambda frame: self.timekeeping_controller.sync_time(int.from_bytes(frame.data, 'big')))
-
-        
+        #TODO: call handle route request when we get a route request
+        #TODO: call handle route response when we get a route response
```
Comment (JonaVDM) [inline libs/Node.py:101]: Uh wut?

File: libs/controllers/config/NodeConfigData.py [hunk 2]
```diff
@@ -41,6 +41,9 @@ def __init__(self, addr: int, measurement_interval: int, replication_count, repl
         self.measurement_interval = measurement_interval
         self.replication_count = replication_count
         self.bidding_wait = bidding_wait
+        self.routing_table: dict[int, tuple[int, int]] = {1: [2,3], 3:[3,2]}    # destination: (next_hop, hops)
+        self.neighbours = {}
+
```
Comment (JonaVDM) [inline libs/controllers/config/NodeConfigData.py:46]: We have a neighbours controller for string this stuff already?

File: libs/controllers/config/NodeConfigData.py [hunk 3]
```diff
@@ -61,6 +64,16 @@ def deserialize(bits):
...
         return NodeConfigData(**umsgpack.loads(bits))
 
+    def add_neighbour(self, addr, node_config):
+        print("adding neighbour")
+        self.neighbours[addr] = node_config
+        print(self.neighbours)
+
+    def remove_neighbour(self, addr):
+        print("removing neighbour")
+        self.neighbours.pop(addr)
```
Comment (JonaVDM) [inline libs/controllers/config/NodeConfigData.py:74]: First off, not the place for this.

Second off, we have a logger now, please use that instead.

File: libs/controllers/neighbours/__init__.py [hunk 4]
```diff
@@ -56,6 +58,9 @@ def handle_join(self, frame: Frame):
 
         node = NodeConfigData.deserialize(frame.data)
         self.connections[node.addr] = node
+        self.node_config.add_neighbour(node.addr, node)
+        print("node joining")
+        print(self.connections)
```
Comment (JonaVDM) [inline libs/controllers/neighbours/__init__.py:63]: Not the place for this.

File: libs/controllers/neighbours/__init__.py [hunk 5]
```diff
@@ -25,9 +25,10 @@ def broadcast_join(self):
         Broadcasts a message with its config for when the node starts up.
         """
         self.network.send_message(
-            Frame.FRAME_TYPES['node_joining'], self.node_config.serialize())
+            Frame.FRAME_TYPES['node_joining'], self.node_config.serialize(), self.node_config.addr)
 
     def start(self):
+        print("starting neighbours controller")
```
Comment (JonaVDM) [inline libs/controllers/neighbours/__init__.py:31]: Why do we need to log this? Also please use the logger.

File: libs/controllers/network/E220NetworkController.py [hunk 6]
```diff
@@ -39,10 +46,90 @@ async def _start(self):
...
+
+        
+        # frame = Frame(type, message, self.address, addr)
+        last_hop = self.address
+        frame = Frame(type, message, source, destination, last_hop)
+        print("frame is", frame.__dict__)
+        print("addr is", addr)
+        if addr == 255:
+            logger(f'sending frame {frame.__dict__}', channel='send_message')
+            self.q.pop() if is_in_queue else None
```
Comment (JonaVDM) [inline libs/controllers/network/E220NetworkController.py:73]: Sorry, uuh, wha

Comment (MichelGerding) [inline libs/controllers/network/E220NetworkController.py:73]: ??????

**Comment by MichelGerding** [comment]
why did you try to run it locally and not remove pycache folders

File: main.py [hunk 7]
```diff
@@ -34,6 +34,8 @@
     replication_count=4
 )
 
+network_controller = E220NetworkController(E220(uart=uart, m0=m0, m1=m1), node_config, set_config=False)
```
Comment (MichelGerding) [inline main.py:37]: there already is a network controller

File: main.py [hunk 8]
```diff
@@ -34,6 +34,8 @@
...
+
```
Comment (JonaVDM) [inline main.py:39]: So now we have two network controllers?

File: libs/helpers/__pycache__/dict.cpython-310.pyc
Comment (JonaVDM) [inline libs/helpers/__pycache__/dict.cpython-310.pyc:1]: Please remove this

File: libs/helpers/__pycache__/__init__.cpython-310.pyc
Comment (JonaVDM) [inline libs/helpers/__pycache__/__init__.cpython-310.pyc:1]: Please remove this

File: libs/external/umsgpack/__pycache__/__init__.cpython-310.pyc
Comment (JonaVDM) [inline libs/external/umsgpack/__pycache__/__init__.cpython-310.pyc:1]: Please remove this

File: libs/external/__pycache__/__init__.cpython-310.pyc
Comment (JonaVDM) [inline libs/external/__pycache__/__init__.cpython-310.pyc:1]: Please remove this

File: libs/controllers/storage/__pycache__/__init__.cpython-310.pyc
Comment (JonaVDM) [inline libs/controllers/storage/__pycache__/__init__.cpython-310.pyc:1]: Please remove this

**Review by JonaVDM — COMMENTED** [review]
First comment, make sure to merge the main branch sometimes.

**Pull request closed without merging**